### PR TITLE
[Store Customization] AI-generated store name

### LINKED ISSUE #11845: Store Customization > Make sure an AI-generated store name is created if the user didn't add any during when creating a new store
This issue is for us to explore adding AI-generated content for the site name if, for some reason, the user didn't set any during the initial onboarding flow:

Screenshot 2023-11-20 at 16 55 32

## Scope

- Find out where this information is saved in the DB
- Ensure that if the business name is empty, AI will automatically generate a new one for the user
- If the user already provided a business description and the relevant option is not empty in the DB, no changes by AI should be made.

### PR DESCRIPTION
**Pull request by albarin**
## What

This PR adds a new endpoint `/ai/store-title` to update the store title during the store customization process if it is empty.

## Why

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/11845

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Setup

Please test the feature on a WooExpress install instead of Jurassic Ninja, as the latter can be very limited in terms of server resources and unstable, as other installs can interfere with your own.

- If you already have a pre-configured install, remove the pre-existing WooCommerce Blocks Plugin and upload the one from this zip file: https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-11878.zip
- Alternatively, create a new WooExpress install from scratch 
- Remove the pre-existing WooCommerce plugin via SSH and
- Install and activate WooCommerce from the following zip file: [woocommerce.zip](https://github.com/woocommerce/woocommerce-blocks/files/13437958/woocommerce.zip)
- Install and activate WooCommerce Blocks from the following zip file: https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-11878.zip
- Install and activate WooCommerce Beta Tester (the plugin is available [within the monorepo](https://github.com/woocommerce/woocommerce/tree/trunk/plugins))
- Go to Settings > General, empty the Site Title and save.
- Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag:

Screenshot 2023-10-22 at 10 32 24

### Test the CYS experience

0. Make sure you don't have any products in your store.
1. Head over to Home > Customize your Store:
 
Screenshot 2023-10-22 at 09 29 08

2. Click on design with AI.
3. Type a description for your business.
4. Finish the customization process and make sure the Site Title shown in the assembler is not empty and it's related to your store business description.
5. Go back to the WooCommerce home and start the process again, make sure the title is not modified this time (because it was not empty from the previous execution).

* [x] Do not include in the Testing Notes 
* [x] Should be tested by the development team exclusively 

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [ ] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

### CODE REVIEW

File: src/StoreApi/Routes/V1/AI/StoreTitle.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	 * @var string
+	 */
+	const STORE_TITLE_OPTION_NAME = 'blogname';
+
+	/**
+	 * The default store title.
+	 *
+	 * @var string
+	 */
+	const DEFAULT_TITLE = 'Site Title';
```
Comment (gigitux) [inline src/StoreApi/Routes/V1/AI/StoreTitle.php:40]: Is the default title the same for all the languages?

File: src/StoreApi/Routes/V1/AI/StoreTitle.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+					'invalid_business_description',
+					__( 'Invalid business description.', 'woo-gutenberg-products-block' )
+				)
+			);
+		}
+
+		$store_title = get_option( 'blogname' );
+		if ( ! ( empty( $store_title ) || self::DEFAULT_TITLE === $store_title ) ) {
+			return rest_ensure_response( array( 'ai_content_generated' => true ) );
+		}
```
Comment (gigitux) [inline src/StoreApi/Routes/V1/AI/StoreTitle.php]: If the default title is defined, should it not be better to update it with AI content?

File: src/StoreApi/Routes/V1/AI/StoreTitle.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+				new \WP_Error(
+					'invalid_business_description',
+					__( 'Invalid business description.', 'woo-gutenberg-products-block' )
+				)
+			);
+		}
+
+		$store_title = get_option( 'blogname' );
+		if ( ! ( empty( $store_title ) || self::DEFAULT_TITLE === $store_title ) ) {
+			return rest_ensure_response( array( 'ai_content_generated' => true ) );
```
Comment (gigitux) [inline src/StoreApi/Routes/V1/AI/StoreTitle.php]: It could make sense to return `false`
```suggestion
			return rest_ensure_response( array( 'ai_content_generated' => false ) );
```

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for working on this!
Overall the PR looks good to me! I left some questions!

File: src/StoreApi/Routes/V1/AI/StoreTitle.php (hunk 1 above)
Comment (albarin) [reply src/StoreApi/Routes/V1/AI/StoreTitle.php:40]: Testing in WooExpress, I haven't found a way to create a new site directly in another language. And when creating a site and switching languages the title remains the same, it's not translated.

File: src/StoreApi/Routes/V1/AI/StoreTitle.php (hunk 2 above)
Comment (albarin) [reply src/StoreApi/Routes/V1/AI/StoreTitle.php]: Yes, that's the idea, the condition is negated.

File: src/StoreApi/Routes/V1/AI/StoreTitle.php (hunk 3 above)
Comment (albarin) [reply src/StoreApi/Routes/V1/AI/StoreTitle.php]: Totally! Updated 🙂

File: src/StoreApi/Routes/V1/AI/StoreTitle.php (hunk 2 above)
Comment (gigitux) [reply src/StoreApi/Routes/V1/AI/StoreTitle.php]: Missed a parenthesis :D

**Review by gigitux — APPROVED** [review]
LGTM! 🚀
